### ISSUE TITLE
Resources text format usage

### ISSUE THREAD

**Issue by arcanis** [godotengine/godot#4420]
I'm using script files to generate some scenes of my projects. It works fine with your text format, except when it comes to asset management. Unless I've missed it, I haven't been able to find documentation about the file format for `tex`, `atex`, `fnt` and similar binary formats, which makes them quite hard to correctly generate. Looking at the issues and the source code, I've been able to find references to a `tres` format, but I'm not sure if it's what I'm looking for, since it doesn't seem to be used for textures and fonts in the example files.

Basically, I'd like to be able to write a file with the following:

```
[texture]
image_path = /assets/directory/myImage.png
image_data = 45A45D21B454BADDCFFDE88==
```

Is that possible, one way or the other?

I'm aware that it would increase the project size and loading time, and that not every resource needs to be imported first (but some are, such as fonts, so why not do it for everything since a script will do it anyway), but I'd still like to know if it's possible. Being able to dynamically generate resources pre-build would be a tremendous help.

**Comment by arcanis** [comment]
I've figured the following from a font resource. However, I'm not sure what this `333?` (0x3FFFFFFF) means, anybody to help?

```
RSRC                   Magic number
^@^@^@^@               Endianness
^@^@^@^@               Is 64 bits file (false)
^A^@^@^@               Version major
^@^@^@^@               Version minor
^@^@^@^@               Format version
^E^@^@^@Font^@         Resource type
\237\276^B^@^@^@^@^@   Offset to impoty metadata
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^@^@^@^@               Reserved
^M^@^@^@               String table size

^N^@^@^@resource/name^@
^F^@^@^@flags^@
^F^@^@^@image^@
^E^@^@^@size^@
^H^@^@^@storage^@
^N^@^@^@lossy_quality^@
^N^@^@^@script/script^@
 ^@^@^@textures^@
^F^@^@^@chars^@
 ^@^@^@kernings^@
^G^@^@^@height^@
^G^@^@^@ascent^@
^O^@^@^@distance_field^@

^@^@^@^@               Amount of external resources
^B^@^@^@               Amount of internal resources
\n^@^@^@local://0^@    Subresource #1 actual path
E^A^@^@^@^@^@^@        Subresource #1 64 bits offset
\n^@^@^@local://1^@    Subresource #2 actual path
\232&^B^@^@^@^@^@      Subresource #2 64 bits offset
^M^@^@^@ImageTexture^@ Subresource #1 type
^E^@^@^@               Subresource #1 properties count
^A^@^@^@               Subresource #1 property #1 name index
^C^@^@^@               Subresource #1 property #1 variant type (float)
^D^@^@^@               Subresource #1 property #1 variant value (float)
^B^@^@^@               Subresource #1 property #2 name index
^U^@^@^@               Subresource #1 property #2 variant type (image)
^B^@^@^@               Subresource #1 property #2 image encoding
\372$^B^@              Subresource #1 property #2 image size
<a lot>                Subresource #1 property #2 image data
<a lot>                Subresource #1 property #2 image padding
^C^@^@^@               Subresource #1 property #3 name index
^B^@^@^@               Subresource #1 property #3 variant type (int)
^E^@^@^@               Subresource #1 property #3 variant value (int)
^D^@^@^@               Subresource #1 property #4 name index
333?
^E^@^@^@Font^@
^F^@^@^@
^G^@^@^@
...
```

**Comment by reduz** [comment]
If you are generating stuff, I'm not sure you want to generate a .tex file manually. Why not just use a png file? Or are you doing it for 3D?

**Comment by reduz** [comment]
RSRC is a compressed resource, it uses a special file compression so you won't be able to simply parse it, you need to zlib-decompress it by blocks. It's not very simple..

**Comment by arcanis** [comment]
I'm doing this for 2D. Basically, I have a PSD file - I'd like to parse it, extract each layer into various pictures, import these textures into Godot, import the needed fonts from the filesystem or an asset project, and generate a scene that uses all of this. The psd-parsing thing is already working great, as is the scene generation. The texture import is fine, but I can't specify any import options on a per-file basis, during the generation (since they're only png files). And as far as the texture import goes, I'm a bit stuck, since generating bmfont files doesn't seem practical.

At first, my idea was to create 'placeholder' files, that would contains the relative path to the source files and a boggus last-update timestamp (01/01/1970-like) so that Godot would automatically "re-import" the source files (actually importing them for the first time) - this way, I would only have to deal with metadata, and let Godot do the heavy work. However, fonts apparently don't support re-import, so once again, I'm a bit stuck :(

**Comment by reduz** [comment]
I wouldn't bother creating binary files.. you can use png files fine

**Comment by arcanis** [comment]
But how about the fonts, or what if I need to set custom export parameters on some textures? On top of that, a few of these textures should belong to the same atlas, so if I can't generate the resource files, I will have to generate image groups ... that doesn't sound very practical :(

**Comment by reduz** [comment]
Godot should now support dynamic fonts, so you can use these. Try save any
of those resources as tres and you should see the text format, should work
for fonts and other stuff.

**Comment by arcanis** [comment]
@reduz I've finally understood how work subresources stored in tres files, and it seems to be very close to what I was looking for! However, is it possible to embed the dynamic font data inside the file? When I tried, I got the following:

```
[gd_scene load_steps=5 format=1]

[sub_resource type="DynamicFontData" id=3]

[sub_resource type="DynamicFont" id=2]

size = 20
font = SubResource( 3 )

```

As you can see, the DynamicDataFont is missing a data property (similar to what `image` is to ImageTexture nodes).

**Comment by arcanis** [comment]
Actually, I've opened a new issue (#4600), and I'm gonna close this one. Thanks for your help!

**Comment by jjay** [comment]
@arcanis Have you any success with importing psd? Sounds like huge job. It could save a lot of time to some of use if you share your experience =)

**Comment by arcanis** [comment]
Hey @jjay - Yes, I've successfully been able to "cast" my psd file into a standalone `.tscn` file ready to be used in Godot. In order to do this, I used a node.js script w/ the `pegjs` package to parse the existing tscn files, the `psd` package to extract layer infos from the psd files, and a custom built library to work with the tscn tree.

My code currently only works on my specific psd file (I hardcoded names etc), but I'd very much like to make a generic utility that could be used on any psd file. Even supporting only a subset of Photoshop's features could probably help quite a few people. I'll probably be able to work on this in the next months.
